python-checks/src/test/resources/checks/inconsistentTypeHintCheck.py

```diff
@@ -0,0 +1,109 @@
+from typing import SupportsFloat, List, Iterable, Generator, Set, Union, Type
+
+def assigned_directly():
+  my_int_nok: int = "hello"  # Noncompliant  {{Assign to "my_int_nok" a value of type "int" instead of "str" or update its type hint.}}
+#             ^^^>  ^^^^^^^
+  my_str_ok: str = 42  # Noncompliant
+  my_int_ok: int = 42  # OK
+  my_str_ok: str = "hello"  # OK
+
+def return_union() -> Union[str, float]:
+  ...
+
+def assigned_to_union(cond):
+  my_int_nok: int = return_union()  # Noncompliant  {{Assign to "my_int_nok" a value of type "int" instead of "Union[str, float]" or update its type hint.}}
+  if cond:
+    x = "hello"
+  else:
+    x = 42.5
+  my_int_nok: int = x  # Noncompliant  {{Assign to "my_int_nok" a value of type "int" or update its type hint.}}
+
+
+def assigned_later(param):
+  a: int
+  a = "hello"  # FN
+  b: int
+  if param:
+    b = 42
+  else:
+    b = "42"  # FN
+
+  c: int
+  c = 1 if a else 2  # OK
+  d: int
+  d = 1 if a else "hello"  # FN
+
+
+def custom_classes():
+  class A:
+    def method(): ...
+
+  class B(A):
+    def additional_method(): ...
+
+  my_a_ok: A = A()  # OK
+  my_a_ok2: A = B()  # OK
+  my_a_nok: A = A  # Noncompliant
+  my_b_nok: B = A()  # Noncompliant {{Assign to "my_b_nok" a value of type "B" instead of "A" or update its type hint.}}
+#           ^>  ^^^
+  my_b_ok: B = B()
+
+
+def get_generator():
+  yield 1
+
+
+def type_aliases():
+  """We should avoid raising FPs on type aliases"""
+  my_float: SupportsFloat = 42  # OK
+  my_iterable: Iterable = []  # OK
+  my_generator: Generator = get_generator()  # OK
+
+
+def collections():
+  my_list: List = {}  # Noncompliant
+
+  my_str_list_nok: List[str] = [1, 2, 3]  # Out of scope
```

Comment:
I think we can mark it as FN, as we'll probably want to detect it when we'll implement MMF-1977